scripts/vulnerabilities.sh

```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
+#
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script prints a table of fixable vulnerabilities of medium severity or
+# higher, based on the latest or specified git tag.
+#
+# USAGE: $ scripts/vulnerabilities.sh [IMAGE_TAG]
+#
+# The IMAGE_TAG is optional. If not specified, it defaults to the latest tag of
+# the current local environment.
+#
+# Requires gcloud authn/authz to read the results of vulnerability scans from
+# Google Container Registry in the "config-management-release" project.
+# Required access:
+# - Storage Object Viewer on "artifacts.config-management-release.appspot.com"
+# - Container Registry Service Agent on Container Registry API
+#   (containerregistry.googleapis.com)
+# - Container Analysis Service Agent on Container Analysis API
+#   (containerscanning.googleapis.com)
+
+set -euo pipefail
+
+configsync_latest_image_tag="$(git describe --tags --abbrev=0)"
+configsync_image_tag="${1:-${configsync_latest_image_tag}}"
+
+# TODO: share with Makefile to keep in sync
+configsync_images=(
+  gcr.io/config-management-release/reconciler
+  gcr.io/config-management-release/reconciler-manager
+  gcr.io/config-management-release/admission-webhook
+  gcr.io/config-management-release/hydration-controller
+  gcr.io/config-management-release/hydration-controller-with-shell
+  gcr.io/config-management-release/oci-sync
+  gcr.io/config-management-release/helm-sync
+  gcr.io/config-management-release/nomos
+)
+
+dependency_images=(
+  gcr.io/config-management-release/resource-group-controller
+  gcr.io/config-management-release/git-sync
+  gcr.io/config-management-release/otelcontribcol
+)
+
+# find the "name:tag" of images in the manifest directory by name
+find_manifest_images_by_name() {
+  image_name="$1"
+  grep -roh --color=never \
+    "${image_name}:\S*" \
+    ./manifests/ | uniq -u
+}
+
+# Build a full list of images with tags
+images=()
+
+# Add the specified tag to all configsync images
+for image_name in "${configsync_images[@]}"; do
```

Comment:
An alternative approach I'm considering is taking in a `config-sync-manifest.yaml` and parsing the images from there. This would make the script work better across versions and help avoid the situation of someone forgetting to update this script (i.e. mismatch of actual images included in a release version versus what is declared in this script). The script could potentially accept either a local file path (e.g. `.output/staging/operator/config-sync-manifest.yaml`) or a GCS path. WDYT?

Reply:
> An alternative approach I'm considering is taking in a config-sync-manifest.yaml and parsing the images from there.

The manifest isn't rendered. It doesn't have the image names in it. We would have to render them first to be able to pull them out of the yaml. I'm already pulling the hardcoded dependencies out of the manifest yaml.

Reply:
One could either target already built manifests published in GCS, or checkout the tagged commit and run the `manifests` target to build the manifests at that SHA. This ensures there isn't a mismatch between what is included in the manifest bundle and what we scan for here

Reply:
Perhaps. I didn't want to make this depend on the manifests being rendered locally. I suppose we could download the published manifests? Or do we only publish the ACM ones?

Reply:
For the CI/CD use case I'm thinking we would pull published manifests from GCS (probably from a bucket in the CI project).

The flexibility to use a manifest on the filesystem would be a convenience for the manual/local use case, IMO. You could either download a published manifest or use manifests that you built locally

Reply:
What if we commit the generated manifest into the repo? We could add a CI check to make sure it's up to date. Then we can just pull the latest from the local path AND we'll be able to download it raw from github for the ACM release process.

Reply:
I'm not sure how practical it would be. Images are typically tagged based on something after the commit is created, e.g. a commit tag or commit SHA - so it becomes a bit of a chicken and egg problem. We could use something arbitrary like a timestamp, but the image tag loses meaning/context. We also would need to choose a registry for these snapshot builds, which begs the question of when/how the images get published. Are they published before the change is submitted, or in a postsubmit job after?

I think the pattern of publishing the manifest and images together after the commit is submitted is a bit more of a natural fit

Reply:
Fair enough. The other option is to publish the rendered manifest to GCS, like we do for ACM.

Reply:
Can we do this later? I don't feel it's urgent to get into this PR.

Reply:
Sure, can be a future improvement